MediaBrowser.MediaEncoding/Encoder/EncoderValidator.cs

```diff
@@ -58,18 +123,127 @@ public bool ValidateVersion(string encoderAppPath, bool logOutput)
+        /// we have stored.
+        /// </summary>
+        /// <param name="output"></param>
+        /// <returns></returns>
+        static private FFmpegVersion GetFFmpegVersion(string output)
+        {
+            // For pre-built binaries the FFmpeg version should be mentioned at the very start of the output
+            var match = Regex.Match(output, @"ffmpeg version (\d+\.\d+)");
+
+            if (match.Success)
+            {
+                return new FFmpegVersion(match.Groups[1].Value);
+            }
+            else
+            {
+                // Try and use the individual library versions to determine a FFmpeg version
+                // This lookup table is to be maintained with the following command line:
+                // $ ./ffmpeg.exe -version | perl -ne ' print "$1=$2.$3," if /^(lib\w+)\s+(\d+)\.\s*(\d+)/'
+                ReadOnlyDictionary<FFmpegVersion, string> lut = new ReadOnlyDictionary<FFmpegVersion, string>
+                    (new Dictionary<FFmpegVersion, string>
+                    {
+                        { new FFmpegVersion("4.1"), "libavutil=56.22,libavcodec=58.35,libavformat=58.20,libavdevice=58.5,libavfilter=7.40,libswscale=5.3,libswresample=3.3,libpostproc=55.3," },
+                        { new FFmpegVersion("4.0"), "libavutil=56.14,libavcodec=58.18,libavformat=58.12,libavdevice=58.3,libavfilter=7.16,libswscale=5.1,libswresample=3.1,libpostproc=55.1," },
+                        { new FFmpegVersion("3.4"), "libavutil=55.78,libavcodec=57.107,libavformat=57.83,libavdevice=57.10,libavfilter=6.107,libswscale=4.8,libswresample=2.9,libpostproc=54.7," },
+                        { new FFmpegVersion("3.3"), "libavutil=55.58,libavcodec=57.89,libavformat=57.71,libavdevice=57.6,libavfilter=6.82,libswscale=4.6,libswresample=2.7,libpostproc=54.5," },
+                        { new FFmpegVersion("3.2"), "libavutil=55.34,libavcodec=57.64,libavformat=57.56,libavdevice=57.1,libavfilter=6.65,libswscale=4.2,libswresample=2.3,libpostproc=54.1," },
+                        { new FFmpegVersion("2.8"), "libavutil=54.31,libavcodec=56.60,libavformat=56.40,libavdevice=56.4,libavfilter=5.40,libswscale=3.1,libswresample=1.2,libpostproc=53.3," }
+                    });
+
+                // Create a reduced version string and lookup key from dictionary
+                var reducedVersion = GetVersionString(output);
+                var found = lut.FirstOrDefault(x => x.Value == reducedVersion).Key;
+
+                if (found != null)
+                {
+                    return found;
+                }
+                else
+                {
+                    // Unknown version.  Test the main libavcoder version in the candidate with the
+                    // latest from the dictionary.  If candidate is greater than dictionary chances are
+                    // the user if running HEAD/master ffmpeg build (which is probably ok)
+                    var firstElement = lut.FirstOrDefault();
```

Comment:
huh, are you sure dictionaries guarantee the order?..

Reply:
I think dictionaries are like Perl hashes, so order is non-deterministic.  But with my latest changes I think this is ok.